### PR TITLE
Add variants to SenderData to represent different verification states

### PR DESCRIPTION
**Pull request by uhoreg**
as [previously discussed](https://github.com/matrix-org/matrix-rust-sdk/pull/3701/files#r1721932382)

- [ ] Public API changes documented in changelogs (optional)

Signed-off-by:

### CODE REVIEW

**Comment by uhoreg** [comment]
One test is known to fail because I need to work on migrating data. And I probably need to write a changelog. But otherwise this PR should be reviewable, and I wanted to get it out so that I'm not holding up Andy's work.

File: crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data.rs
Comment (andybalaam) [inline crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data.rs:23]: their->the?

File: crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data.rs
Comment (andybalaam) [inline crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data.rs:77]: send->sent

File: crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data.rs
Comment (andybalaam) [inline crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data.rs:168]: It looks like this comment was wrong before, but definitely needs updating now.

**Review by andybalaam — APPROVED** [review]
Looks good to me. My only question is whether the already-incorrect doc comment on `is_known` was actually describing the behaviour some other code expected.

We also need to think about migration, as you mention.

File: crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data.rs
Comment (uhoreg) [reply crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data.rs:169]: Looking at this function, it looks like it's only used in OlmMachine to determine whether it should recalculate the sender_data.  So I definitely don't want it to return `true` if it's `SenderUnverifiedButPreviouslyVerified`.  But I'm not sure that it should return `true` if it's `SenderUnverified` either, although that would match the previous behaviour, since `sender_data_to_verification_state` returns different values if the sender is verified or not.  What do you think?  What was your rationale for having it return true when the sender was unverified?

Comment (andybalaam) [reply crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data.rs:169]: I'm happy for you to merge this as-is or with one of the changes I mention above. If you want to change the behaviour I'll want to discuss it further.

Comment (uhoreg) [reply crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data.rs]: > Maybe this function should be moved inside `OlmMachine` and be named `should_recalculate`?

Done.  It's ended up adding a bunch of negations, which I'm not really a fan of, but to avoid the negations, I'd either need to rename to `should_not_recalculate`, or switch the logic in the place where it's called, both of which seem at least as bad.

If you're OK with the way it is, can you hit the merge button?  I don't have permissions.

**Comment by andybalaam** [comment]
I wanted to make some very minor changes, so closing in favour of https://github.com/matrix-org/matrix-rust-sdk/pull/3877

**Pull request closed without merging**
